[core IR] Add decomps for `aten.sum` and `aten.squeeze` variants

SS-JIA opened the pull request:
Summary:
## Context

Both `aten.sum` and `aten.squeeze` have a "most generic" variant in the form of `aten.sum.dim_IntList` and `aten.squeeze.dims` respectively. Add decompositions for other non generic variants of these operators to express them using the most generic variant.

Note that to register these decomps, the reference implementation under `_refs` had to be removed as registered decompositions. cc: @lezcano @peterbell10 

Test Plan: Github CI + Meta Internal CI

Differential Revision: D49965952




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

peterbell10 commented on torch/_refs/__init__.py:
```diff
@@ -3987,7 +3986,6 @@ def index_select(x: TensorLike, dim: int, index: TensorLike):
     return x[idx]
 
 
-@register_decomposition(aten.squeeze)
```
We can still register these for `sum,dim_IntList` and `squeeze.dims` which should fix the decomp test.

SS-JIA replied:
ah, good point. Let me update.

peterbell10 commented on test/test_decomp.py:
```diff
@@ -382,6 +382,12 @@ def normalize_op_input_output(f, sample, requires_grad=True):
+SKIP_CHECK_DECOMPOSED = {
+    "sum",
+    "squeeze",
+}
```
Can we revert these skips?

SS-JIA replied:
Yep, just did it! Forgot to remove these the first update.

SS-JIA commented:
Address failures + consolidate code

SS-JIA commented:
Seems all checks are passing; @peterbell10 let me know if there are any additional concerns. Thanks in advance!

lezcano commented on torch/_decomp/decompositions.py:
```diff
@@ -4211,6 +4211,28 @@ def floor_divide(self, other):
+        return aten.sum.IntList_out(self, [], keepdim=keepdim, dtype=dtype, out=out)
+
+
+@register_decomposition([aten.squeeze.default, aten.squeeze.dim])
+def squeeze_default(self: Tensor, dim: Optional[int] = None):
+    if dim is None:
+        return aten.squeeze.dims(self, list(range(self.dim())))
+    else:
+        return aten.squeeze.dims(self, [dim])
```
Isn't this case already handled in the reference for squeeze?

SS-JIA replied:
The reference implementation for squeeze calls `prims.squeeze`; for core decompositions, we want to avoid decomposing into prim ops (i.e. only decompose to other ATen operators). Hence why a new decomp was added instead of using the reference impl directly.

lezcano commented on torch/_decomp/decompositions.py:
```diff
@@ -4211,6 +4211,28 @@ def floor_divide(self, other):
 
 
+@register_decomposition(aten.sum.default)
+def sum_default(
+    self: Tensor,
```
How is it possible to reach this overload? It seems like it's not exposed to Python:
```python
>>> torch.randn(3,3).sum(True)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: sum() received an invalid combination of arguments - got (bool), but expected one of:
 * (*, torch.dtype dtype)
 * (tuple of ints dim, bool keepdim, *, torch.dtype dtype)
 * (tuple of names dim, bool keepdim, *, torch.dtype dtype)
```

SS-JIA replied:
Ah, good point. I will update the signature of this function. I saw `keepdim` as an argument and assumed that it could be a valid combination with `dtype` 😅

peterbell10 commented on test/expect/HasDecompTest.test_has_decomposition.expect:
```diff
@@ -1267,6 +1267,8 @@ aten::squeeze_copy.dims
 aten::squeeze_copy.out
 aten::sspaddmm.out
+aten::sum.IntList_out
+aten::sum.out
```
We can register these on the reference as well

SS-JIA replied:
Ah, yes that's true. Will update that. But for `sum.out`, I think I will register to the new decomposition since it can be expressed with `sum.IntList_out` as well.

SS-JIA commented:
Updated to address comments.

lezcano reviewed (commented):
These two were already being decomposed by the ref. It's a bit annoying to have to factor them out into their own decompositions. I think a proper fix going forward would be to fix the issues with prims: Either make the usable (removing the broadcasting behaviour and dispatching them to the core ATen ops) or remove them completely. Otherwise, we are going to end up with plenty of these tiny decompositions, rather than the tighter all-in-one that we have implemented in refs.

cc @ezyang

Commit:
[core IR] Add decomps for `aten.sum` and `aten.squeeze` variants (#110645)

Summary:

## Context

Both `aten.sum` and `aten.squeeze` have a "most generic" variant in the form of `aten.sum.dim_IntList` and `aten.squeeze.dims` respectively. Add decompositions for other non generic variants of these operators to express them using the most generic variant.

Note that to register these decomps, the reference implementation under `_refs` had to be removed as registered decompositions.

Test Plan: Github CI + Meta Internal CI

Reviewed By: manuelcandales

Differential Revision: D49965952
